zerver/views/auth.py

```diff
@@ -855,6 +855,61 @@ def view(request, *args, **kwargs):
     return two_fa_view(request, **kwargs)
 
 
+@csrf_exempt
+@has_request_variables
+def jwt_fetch_api_key(request: HttpRequest, json_web_token: str = REQ(default="")) -> HttpResponse:
+    realm = get_realm_from_request(request)
+    if realm is None:
+        raise JsonableError(_("Invalid subdomain"))
+
+    host = realm.host
+    try:
+        key = settings.JWT_FETCH_API_KEYS[host]["key"]
+        algorithms = settings.JWT_FETCH_API_KEYS[host]["algorithms"]
+    except KeyError:
+        raise JsonableError(_("Auth key for this subdomain not found"))
+
+    try:
+        json_web_token = request.POST["json_web_token"]
+        options = {"verify_signature": True}
+        payload = jwt.decode(json_web_token, key, algorithms=algorithms, options=options)
+    except KeyError:
+        raise JsonableError(_("No JSON web token passed in request"))
+    except jwt.InvalidTokenError:
+        raise JsonableError(_("Bad JSON web token"))
+
+    remote_email = payload.get("email", None)
+    if remote_email is None or not remote_email:
+        raise JsonableError(_("No email specified in JSON web token claims"))
+
+    return_data: Dict[str, bool] = {}
+    user_profile = authenticate(
+        username=remote_email, realm=realm, return_data=return_data, use_dummy_backend=True
+    )
+    if return_data.get("inactive_user"):
+        raise JsonableError(_("Account is deactivated"))
+    if return_data.get("inactive_realm"):
+        raise JsonableError(_("This organization has been deactivated"))
+    if return_data.get("invalid_subdomain"):
+        raise JsonableError(_("Invalid subdomain"))
+    if user_profile is None:
+        raise JsonableError(_("User not found"))
+
+    process_client(request, user_profile)
+
+    api_key = get_api_key(user_profile)
+    members = get_raw_user_data(
+        realm,
+        user_profile,
+        target_user=user_profile,
+        client_gravatar=False,
+        user_avatar_url_field_optional=False,
+        include_custom_profile_fields=False,
+    )
+
+    return json_success(request, data={"api_key": api_key, "user": members[user_profile.id]})
```

Comment:
What's the thinking around the format of this response? I assume you need the `user` object here primarily to send the `delivery_email` to use alongside the API key?

Reply:
I think copying the format of `api_fetch_api_key` would probably be cleaner, if there isn't a strong need for the other data that this would add to what is returned. (I imagine at most you want the `full_name`?)

Reply:
We actually need the whole `user` object as it represents the abstraction of a Zulip user in our system. We might retrieve that info via API by knowing the `user_id`, but we'd rather avoid a further API call (which we'd need to perform for each and every user).

Reply:
Started a new topic here:

https://chat.zulip.org/#narrow/stream/378-api-design/topic/Request.20user.20details.20when.20fetching.20api.20keys.20through.20JWT